src/lxc/console.c

```diff
@@ -124,46 +133,60 @@ struct lxc_tty_state *lxc_console_signal_init(int srcfd, int dstfd)
 	ts->masterfd = dstfd;
 	ts->sigfd = -1;
 
+	sigemptyset(&mask);
+
 	istty = isatty(srcfd) == 1;
 	if (!istty) {
 		INFO("fd %d does not refer to a tty device", srcfd);
-		return ts;
+	} else {
+		/* Add tty to list to be scanned at SIGWINCH time. */
+		lxc_list_add_elem(&ts->node, ts);
+		lxc_list_add_tail(&lxc_ttys, &ts->node);
+		sigaddset(&mask, SIGWINCH);
 	}
 
-	/* add tty to list to be scanned at SIGWINCH time */
-	lxc_list_add_elem(&ts->node, ts);
-	lxc_list_add_tail(&lxc_ttys, &ts->node);
+	/* Exit the mainloop cleanly on SIGTERM. */
+	sigaddset(&mask, SIGTERM);
 
-	sigemptyset(&mask);
-	sigaddset(&mask, SIGWINCH);
-	if (sigprocmask(SIG_BLOCK, &mask, &ts->oldmask)) {
-		SYSERROR("Failed to block SIGWINCH signal");
-		ts->sigfd = -1;
-		lxc_list_del(&ts->node);
-		return ts;
+	ret = sigprocmask(SIG_BLOCK, &mask, &ts->oldmask);
+	if (ret < 0) {
+		WARN("Failed to block signals");
+		goto on_error;
 	}
 
 	ts->sigfd = signalfd(-1, &mask, 0);
 	if (ts->sigfd < 0) {
-		SYSERROR("Failed to create signal fd");
+		WARN("Failed to create signal fd");
 		sigprocmask(SIG_SETMASK, &ts->oldmask, NULL);
-		ts->sigfd = -1;
-		lxc_list_del(&ts->node);
-		return ts;
+		goto on_error;
 	}
 
-	DEBUG("Process %d created signal fd %d", getpid(), ts->sigfd);
+	DEBUG("Created signal fd %d", ts->sigfd);
+	return ts;
+
+on_error:
+	ERROR("Failed to create signal fd");
+	if (ts->sigfd >= 0) {
+		close(ts->sigfd);
+		ts->sigfd = -1;
+	}
+	if (istty)
+		lxc_list_del(&ts->node);
 	return ts;
 }
 
 void lxc_console_signal_fini(struct lxc_tty_state *ts)
 {
 	if (ts->sigfd >= 0) {
 		close(ts->sigfd);
-		lxc_list_del(&ts->node);
-		sigprocmask(SIG_SETMASK, &ts->oldmask, NULL);
+
+		if (sigprocmask(SIG_SETMASK, &ts->oldmask, NULL) < 0)
+			WARN("%s - Failed to restore signal mask", strerror(errno));
 	}
 
+	if (isatty(ts->stdinfd))
```

Comment:
Seems fine.  I would have liked for the patch description to mention that the ts->node will now only be added to if stdinfd is a tty.